File: MyCustomersApp/Services/AuthService.cs
using System.Net.Http.Json;
using System.Text.Json;
using Microsoft.JSInterop;
using MyCustomersApp.Models;

namespace MyCustomersApp.Services;

public class AuthService : IAuthService
{
    private readonly HttpClient _httpClient;
    private readonly HttpClient _publicHttpClient;
    private readonly IJSRuntime _jsRuntime;
    private readonly ILogger<AuthService> _logger;
    private UserInfo? _currentUser;

    public event Action<bool>? AuthStateChanged;

    public AuthService(IHttpClientFactory httpClientFactory, IJSRuntime jsRuntime, ILogger<AuthService> logger)
    {
        _httpClient = httpClientFactory.CreateClient("ApiClient");
        _publicHttpClient = httpClientFactory.CreateClient("PublicApiClient");
        _jsRuntime = jsRuntime;
        _logger = logger;
    }

    public async Task<AuthResponse> LoginAsync(LoginRequest request)
    {
        try
        {
            var identityRequest = new
            {
                email = request.Email,
                password = request.Password
            };
            
            var response = await _publicHttpClient.PostAsJsonAsync("api/identity/login", identityRequest);
            
            if (response.IsSuccessStatusCode)
            {
                var result = await response.Content.ReadFromJsonAsync<JsonElement>();
                var accessToken = result.GetProperty("accessToken").GetString();
                
                if (!string.IsNullOrEmpty(accessToken))
                {
                    await _jsRuntime.InvokeVoidAsync("localStorage.setItem", "authToken", accessToken);
                    
                    // Get user details including admin status from the API
                    var userDetails = await GetUserDetailsAsync(accessToken);
                    
                    await _jsRuntime.InvokeVoidAsync("localStorage.setItem", "userInfo", JsonSerializer.Serialize(userDetails));
                    _currentUser = userDetails;
                    AuthStateChanged?.Invoke(true);
                    
                    return new AuthResponse 
                    { 
                        Success = true, 
                        Token = accessToken,
                        User = userDetails 
                    };
                }
            }
            
            var errorContent = await response.Content.ReadAsStringAsync();
            return new AuthResponse { Success = false, Message = "Login failed: " + errorContent };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error during login");
            return new AuthResponse { Success = false, Message = "Login failed" };
        }
    }

    public async Task<AuthResponse> RegisterAsync(RegisterRequest request)
    {
        try
        {
            // Create Identity API compatible request
            var identityRequest = new
            {
                email = request.Email,
                password = request.Password
            };
            
            var response = await _publicHttpClient.PostAsJsonAsync("api/identity/register", identityRequest);
            
            if (response.IsSuccessStatusCode)
            {
                // Identity API returns different format, create our own response
                var user = new UserInfo
                {
                    Id = Guid.NewGuid().ToString(),
                    Email = request.Email,
                    IsAdmin = false
                };
                
                // Store registered user in localStorage for user management
                await StoreRegisteredUser(user);
                
                // For now, we'll consider registration successful but won't have a token
                // The user will need to log in separately
                return new AuthResponse 
                { 
                    Success = true, 
                    Message = "Registration successful. Please log in.",
                    User = user
                };
            }
            else
            {
                var errorContent = await response.Content.ReadAsStringAsync();
                return new AuthResponse { Success = false, Message = "Registration failed: " + errorContent };
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error during registration");
            return new AuthResponse { Success = false, Message = "Registration failed" };
        }
    }

    public async Task LogoutAsync()
    {
        await _jsRuntime.InvokeVoidAsync("localStorage.removeItem", "authToken");
        await _jsRuntime.InvokeVoidAsync("localStorage.removeItem", "userInfo");
        _currentUser = null;
        AuthStateChanged?.Invoke(false);
    }

    public async Task<bool> IsAuthenticatedAsync()
    {
        try
        {
            var token = await _jsRuntime.InvokeAsync<string>("localStorage.getItem", "authToken");
            return !string.IsNullOrEmpty(token);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error checking authentication status");
            return false;
        }
    }

    public async Task<UserInfo?> GetCurrentUserAsync()
    {
        if (_currentUser != null) return _currentUser;
        
        try
        {
            var userInfoJson = await _jsRuntime.InvokeAsync<string>("localStorage.getItem", "userInfo");
            if (!string.IsNullOrEmpty(userInfoJson))
            {
                _currentUser = JsonSerializer.Deserialize<UserInfo>(userInfoJson);
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting current user");
        }
        
        return _currentUser;
    }

    private async Task<UserInfo> GetUserDetailsAsync(string accessToken)
    {
        try
        {
            // Create a temporary HttpClient with the auth token to get user details
            using var tempClient = new HttpClient();
            tempClient.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", accessToken);
            tempClient.BaseAddress = new Uri(_httpClient.BaseAddress!.ToString());
            
            var response = await tempClient.GetAsync("api/users/me");
            if (response.IsSuccessStatusCode)
            {
                var userDto = await response.Content.ReadFromJsonAsync<UserDto>();
                if (userDto != null)
                {
                    return new UserInfo
                    {
                        Id = userDto.Id,
                        Email = userDto.Email,
                        IsAdmin = userDto.IsAdmin
                    };
                }
            }
            else
            {
                _logger.LogWarning("Failed to get user details from API, using fallback. Status: {StatusCode}", response.StatusCode);
                var errorContent = await response.Content.ReadAsStringAsync();
                _logger.LogWarning("Error content: {ErrorContent}", errorContent);
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting user details from API, using fallback");
        }
        
        // Fallback: create basic user info without admin status
        return new UserInfo
        {
            Id = Guid.NewGuid().ToString(),
            Email = "unknown@example.com",
            IsAdmin = false
        };
    }

    private async Task StoreRegisteredUser(UserInfo user)
    {
        try
        {
            var registeredUsersJson = await _jsRuntime.InvokeAsync<string>("localStorage.getItem", "registeredUsers");
            var registeredUsers = new List<UserInfo>();
            
            if (!string.IsNullOrEmpty(registeredUsersJson))
            {
                registeredUsers = JsonSerializer.Deserialize<List<UserInfo>>(registeredUsersJson) ?? new List<UserInfo>();
            }
            
            // Add user if not already exists
            if (!registeredUsers.Any(u => u.Email!.Equals(user.Email, StringComparison.OrdinalIgnoreCase)))
            {
                registeredUsers.Add(user);
                var updatedJson = JsonSerializer.Serialize(registeredUsers);
                await _jsRuntime.InvokeVoidAsync("localStorage.setItem", "registeredUsers", updatedJson);
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error storing registered user");
        }
    }
}

File: MyCustomersRN/src/services/authService.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { LoginRequest, RegisterRequest, AuthResponse, UserInfo } from '../types/Auth';
import { apiClient } from './apiClient';

export class AuthService {
  private static instance: AuthService;
  private currentUser: UserInfo | null = null;
  private authListeners: ((isAuthenticated: boolean, user: UserInfo | null) => void)[] = [];

  static getInstance(): AuthService {
    if (!AuthService.instance) {
      AuthService.instance = new AuthService();
    }
    return AuthService.instance;
  }

  // Add auth state change listener
  addAuthListener(listener: (isAuthenticated: boolean, user: UserInfo | null) => void) {
    this.authListeners.push(listener);
  }

  // Remove auth state change listener
  removeAuthListener(listener: (isAuthenticated: boolean, user: UserInfo | null) => void) {
    this.authListeners = this.authListeners.filter(l => l !== listener);
  }

  // Notify all listeners of auth state change
  private notifyAuthChange() {
    const isAuthenticated = this.currentUser !== null;
    this.authListeners.forEach(listener => listener(isAuthenticated, this.currentUser));
  }

  async login(email: string, password: string): Promise<AuthResponse> {
    try {
      const loginData = { email, password };
      
      const response = await fetch(`${apiClient.baseURL}/api/identity/login`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(loginData),
      });

      if (response.ok) {
        const result = await response.json();
        const accessToken = result.accessToken;
        
        if (accessToken) {
          const user: UserInfo = {
            id: result.id || new Date().getTime().toString(),
            email: email,
            firstName: result.firstName || '',
            lastName: result.lastName || '',
          };

          // Store token and user info
          await AsyncStorage.setItem('authToken', accessToken);
          await AsyncStorage.setItem('userInfo', JSON.stringify(user));
          
          this.currentUser = user;
          this.notifyAuthChange();

          return {
            success: true,
            token: accessToken,
            user: user,
          };
        }
      }

      const errorText = await response.text();
      return {
        success: false,
        message: `Login failed: ${errorText}`,
      };
    } catch (error) {
      console.error('Login error:', error);
      return {
        success: false,
        message: 'Login failed due to network error',
      };
    }
  }

  async register(userData: Omit<RegisterRequest, 'confirmPassword'>): Promise<AuthResponse> {
    try {
      const registerData = {
        email: userData.email,
        password: userData.password,
      };

      const response = await fetch(`${apiClient.baseURL}/api/identity/register`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(registerData),
      });

      if (response.ok) {
        const user: UserInfo = {
          id: new Date().getTime().toString(),
          firstName: userData.firstName,
          lastName: userData.lastName,
          email: userData.email,
        };

        // Store registered user info for later use
        await this.storeRegisteredUser(user);

        return {
          success: true,
          message: 'Registration successful. Please log in.',
          user: user,
        };
      }

      const errorText = await response.text();
      return {
        success: false,
        message: `Registration failed: ${errorText}`,
      };
    } catch (error) {
      console.error('Registration error:', error);
      return {
        success: false,
        message: 'Registration failed due to network error',
      };
    }
  }

  async logout(): Promise<void> {
    try {
      await AsyncStorage.removeItem('authToken');
      await AsyncStorage.removeItem('userInfo');
      this.currentUser = null;
      this.notifyAuthChange();
    } catch (error) {
      console.error('Logout error:', error);
    }
  }

  async isAuthenticated(): Promise<boolean> {
    try {
      const token = await AsyncStorage.getItem('authToken');
      return !!token;
    } catch (error) {
      console.error('Error checking authentication:', error);
      return false;
    }
  }

  async getCurrentUser(): Promise<UserInfo | null> {
    if (this.currentUser) {
      return this.currentUser;
    }

    try {
      const userInfoJson = await AsyncStorage.getItem('userInfo');
      if (userInfoJson) {
        this.currentUser = JSON.parse(userInfoJson);
      }
    } catch (error) {
      console.error('Error getting current user:', error);
    }

    return this.currentUser;
  }

  async getAuthToken(): Promise<string | null> {
    try {
      return await AsyncStorage.getItem('authToken');
    } catch (error) {
      console.error('Error getting auth token:', error);
      return null;
    }
  }

  private async storeRegisteredUser(user: UserInfo): Promise<void> {
    try {
      const registeredUsersJson = await AsyncStorage.getItem('registeredUsers');
      let registeredUsers: UserInfo[] = [];

      if (registeredUsersJson) {
        registeredUsers = JSON.parse(registeredUsersJson);
      }

      // Add user if not already exists
      const existingUser = registeredUsers.find(u => 
        u.email.toLowerCase() === user.email.toLowerCase()
      );

      if (!existingUser) {
        registeredUsers.push(user);
        await AsyncStorage.setItem('registeredUsers', JSON.stringify(registeredUsers));
      }
    } catch (error) {
      console.error('Error storing registered user:', error);
    }
  }
}

export const authService = AuthService.getInstance();